src/Functions/serial.cpp

```diff
@@ -0,0 +1,161 @@
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int KEEPER_EXCEPTION;
+}
+
+class FunctionSerial : public IFunction
+{
+private:
+    mutable zkutil::ZooKeeperPtr zk{nullptr};
+    ContextPtr context;
+
+public:
+    static constexpr auto name = "serial";
+
+    explicit FunctionSerial(ContextPtr ctx) : context(ctx)
+    {
+        if (ctx->hasZooKeeper()) {
+            zk = ctx->getZooKeeper();
+        }
+    }
+
+    static FunctionPtr create(ContextPtr context)
+    {
+        return std::make_shared<FunctionSerial>(std::move(context));
+    }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool isStateful() const override { return true; }
+    bool isDeterministic() const override { return false; }
+    bool isDeterministicInScopeOfQuery() const override { return false; }
+    bool isSuitableForConstantFolding() const override { return false; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+    bool useDefaultImplementationForNothing() const override { return false; }
+    bool canBeExecutedOnDefaultArguments() const override { return false; }
+    bool isInjective(const ColumnsWithTypeAndName & /*sample_columns*/) const override { return true; }
+    bool hasInformationAboutMonotonicity() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() != 1)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Number of arguments for function {} doesn't match: passed {}, should be 1.",
+                getName(), arguments.size());
+        if (!isStringOrFixedString(arguments[0]))
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Type of argument for function {} doesn't match: passed {}, should be string",
+                getName(), arguments[0]->getName());
+
+        return std::make_shared<DataTypeInt64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        if (zk == nullptr)
+            throw Exception(ErrorCodes::KEEPER_EXCEPTION,
+            "ZooKeeper is not configured for function {}",
+            getName());
+        if (zk->expired())
+            zk = context->getZooKeeper();
+
+        auto col_res = ColumnVector<Int64>::create();
+        typename ColumnVector<Int64>::Container & vec_to = col_res->getData();
+        size_t size = input_rows_count;
+        vec_to.resize(size);
+
+        const auto & serial_path = "/serials/" + arguments[0].column->getDataAt(0).toString();
```

Comment:
"serials" sounds a bit odd, maybe "serial_ids"?

Reply:
So if the argument of `serial` is non-const (which is the default case), we will always use the value in the first row as value?

Reply:
Argument should be constant string

Reply:
... which is currently not guaranteed. The place to assert that would be in `getReturnTypeImpl`. This kind of check is actually very easy to do using `FunctionArgumentHelper` (see my other comment).